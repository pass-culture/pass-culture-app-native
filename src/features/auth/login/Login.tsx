import { t } from '@lingui/macro'
import { useNavigation } from '@react-navigation/native'
import React, { useRef, FunctionComponent, useState, useEffect } from 'react'
import { Keyboard, TouchableOpacity } from 'react-native'
import { useQuery } from 'react-query'
import styled from 'styled-components/native'

import { useSignIn } from 'features/auth/api'
import { useBackNavigation } from 'features/navigation/backNavigation'
import { NavigateToHomeWithoutModalOptions } from 'features/navigation/helpers'
import { UseNavigationType } from 'features/navigation/RootNavigator'
import { env } from 'libs/environment'
import { _ } from 'libs/i18n'
import { BottomContentPage } from 'ui/components/BottomContentPage'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { isValueEmpty } from 'ui/components/inputs/helpers'
import { InputError } from 'ui/components/inputs/InputError'
import { PasswordInput } from 'ui/components/inputs/PasswordInput'
import { TextInput } from 'ui/components/inputs/TextInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { getSpacing, Spacer, Typo } from 'ui/theme'

let INITIAL_IDENTIFIER = ''
let INITIAL_PASSWORD = ''

if (__DEV__) {
  INITIAL_IDENTIFIER = env.SIGNIN_IDENTIFIER
  INITIAL_PASSWORD = env.SIGNIN_PASSWORD
}

export const Login: FunctionComponent = function () {
  const isMounted = useRef(false)

  const [email, setEmail] = useState(INITIAL_IDENTIFIER)
  const [password, setPassword] = useState(INITIAL_PASSWORD)
  const [shouldShowErrorMessage, setShouldShowErrorMessage] = useState(false)
  const signIn = useSignIn()

  const shouldDisableLoginButton = isValueEmpty(email) || isValueEmpty(password)

  const { navigate } = useNavigation<UseNavigationType>()
  const complexGoBack = useBackNavigation<'Login'>()

  useEffect(() => {
    isMounted.current = true
    return () => {
      isMounted.current = false
    }
  }, [])

  const signInQuery = async () => {
    const signinResponse = await signIn({ identifier: email, password })
    if (signinResponse?.isSuccess) {
      navigate('Home', NavigateToHomeWithoutModalOptions)
    } else {
      const { code } = signinResponse?.content || {}
      if (code === 'EMAIL_NOT_VALIDATED') {
        navigate('SignupConfirmationEmailSent', { email })
      } else if (code === 'NETWORK_REQUEST_FAILED') {
        throw new Error('Offline')
      }
    }
    return signinResponse
  }

  const { refetch, isFetching } = useQuery('login', signInQuery, {
    cacheTime: 0,
    enabled: false,
  })

  async function handleSignin() {
    Keyboard.dismiss()
    setShouldShowErrorMessage(false)
    const { data } = await refetch()
    if (isMounted.current && !data?.isSuccess) {
      setShouldShowErrorMessage(true)
    }
  }

  function onClose() {
    navigate('Home', NavigateToHomeWithoutModalOptions)
  }

  function onForgottenPasswordClick() {
    navigate('ForgottenPassword')
  }

  return (
    <BottomContentPage>
      <ModalHeader
        title={_(t`Connecte-toi !`)}
        leftIcon={ArrowPrevious}
        onLeftIconPress={complexGoBack}
        rightIcon={Close}
        onRightIconPress={onClose}
      />
      <InputError
        visible={shouldShowErrorMessage}
        messageId="E-mail ou mot de passe incorrect"
        numberOfSpacesTop={5}
      />
      <Spacer.Column numberOfSpaces={7} />
      <StyledInput>
        <Typo.Body>{_(t`Adresse e-mail`)}</Typo.Body>
        <Spacer.Column numberOfSpaces={2} />
        <TextInput
          autoCapitalize="none"
          autoFocus={true}
          isError={shouldShowErrorMessage}
          keyboardType="email-address"
          onChangeText={setEmail}
          placeholder={_(/*i18n: email placeholder */ t`tonadresse@email.com`)}
          textContentType="emailAddress"
          value={email}
        />
      </StyledInput>
      <Spacer.Column numberOfSpaces={6} />
      <StyledInput>
        <Typo.Body>{_(t`Mot de passe`)}</Typo.Body>
        <Spacer.Column numberOfSpaces={2} />
        <PasswordInput
          value={password}
          onChangeText={setPassword}
          placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
          isError={shouldShowErrorMessage}
          textContentType="password"
        />
      </StyledInput>
      <Spacer.Column numberOfSpaces={7} />
      <ForgottenPasswordContainer>
        <TouchableOpacity onPress={onForgottenPasswordClick}>
          <Typo.ButtonText>{_(t`Mot de passe oubli√© ?`)}</Typo.ButtonText>
        </TouchableOpacity>
      </ForgottenPasswordContainer>
      <Spacer.Column numberOfSpaces={8} />
      <ButtonPrimary
        title={_(t`Se connecter`)}
        onPress={handleSignin}
        disabled={shouldDisableLoginButton || isFetching}
      />
    </BottomContentPage>
  )
}

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})

const ForgottenPasswordContainer = styled.View({
  flexDirection: 'row',
  justifyContent: 'flex-end',
  width: '100%',
  maxWidth: getSpacing(125),
})
