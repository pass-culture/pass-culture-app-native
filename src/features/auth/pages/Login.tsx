import { t } from '@lingui/macro'
import { StackScreenProps } from '@react-navigation/stack'
import React, { FunctionComponent, useState } from 'react'
import { Alert } from 'react-native'
import styled from 'styled-components/native'

import { RootStackParamList } from 'features/navigation/RootNavigator'
import { env } from 'libs/environment'
import { _ } from 'libs/i18n'
import { BottomCard } from 'ui/components/BottomCard'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { isValueEmpty } from 'ui/components/inputs/helpers'
import { PasswordInput } from 'ui/components/inputs/PasswordInput'
import { TextInput } from 'ui/components/inputs/TextInput'
import { ModalHeader } from 'ui/components/modals/ModalHeader'
import { SafeContainer } from 'ui/components/SafeContainer'
import { ArrowPrevious } from 'ui/svg/icons/ArrowPrevious'
import { Close } from 'ui/svg/icons/Close'
import { Warning } from 'ui/svg/icons/Warning'
import { ColorsEnum, getSpacing, Spacer, Typo } from 'ui/theme'

import { signin } from '../api'

type Props = StackScreenProps<RootStackParamList, 'Login'>

let INITIAL_IDENTIFIER = ''
let INITIAL_PASSWORD = ''

if (__DEV__) {
  INITIAL_IDENTIFIER = env.SIGNIN_IDENTIFIER
  INITIAL_PASSWORD = env.SIGNIN_PASSWORD
}

export const Login: FunctionComponent<Props> = function (props: Props) {
  const [email, setEmail] = useState(INITIAL_IDENTIFIER)
  const [password, setPassword] = useState(INITIAL_PASSWORD)
  const [shouldShowErrorMessage, setShouldShowErrorMessage] = useState(false)

  const shouldDisableLoginButton = isValueEmpty(email) || isValueEmpty(password)

  async function handleSignin() {
    setShouldShowErrorMessage(false)
    const isSigninSuccessful = await signin({ email, password })
    if (isSigninSuccessful) {
      props.navigation.navigate('Home', { shouldDisplayLoginModal: false })
    } else {
      setShouldShowErrorMessage(true)
    }
  }

  function onBackNavigation() {
    props.navigation.navigate('Home', { shouldDisplayLoginModal: true })
  }

  function onClose() {
    props.navigation.navigate('Home', { shouldDisplayLoginModal: false })
  }

  function onForgottenPasswordClick() {
    Alert.alert('TO DO')
  }

  return (
    <SafeContainer>
      <BottomCard>
        <ModalHeader
          title={_(t`Connecte-toi !`)}
          leftIcon={ArrowPrevious}
          onLeftIconPress={onBackNavigation}
          rightIcon={Close}
          onRightIconPress={onClose}
        />
        {shouldShowErrorMessage && (
          <React.Fragment>
            <Spacer.Column numberOfSpaces={5} />
            <StyledInline>
              <Warning size={24} />
              <Typo.Caption color={ColorsEnum.ERROR}>
                {_(t`E-mail ou mot de passe incorrect`)}
              </Typo.Caption>
            </StyledInline>
          </React.Fragment>
        )}
        <Spacer.Column numberOfSpaces={7} />
        <StyledInput>
          <Typo.Body>{_(t`Adresse e-mail`)}</Typo.Body>
          <Spacer.Column numberOfSpaces={2} />
          <TextInput
            value={email}
            onChangeText={setEmail}
            placeholder={_(/*i18n: email placeholder */ t`tonadresse@email.com`)}
            keyboardType="email-address"
            autoFocus={true}
            isError={shouldShowErrorMessage}
            width="100%"
          />
        </StyledInput>
        <Spacer.Column numberOfSpaces={6} />
        <StyledInput>
          <Typo.Body>{_(t`Mot de passe`)}</Typo.Body>
          <Spacer.Column numberOfSpaces={2} />
          <PasswordInput
            value={password}
            onChangeText={setPassword}
            placeholder={_(/*i18n: password placeholder */ t`Ton mot de passe`)}
            isError={shouldShowErrorMessage}
            width="100%"
          />
        </StyledInput>
        <Spacer.Column numberOfSpaces={7} />
        <StyledForgottenPasswordTouchableOpacity onPress={onForgottenPasswordClick}>
          <Typo.ButtonText>{_(t`Mot de passe oubli√© ?`)}</Typo.ButtonText>
        </StyledForgottenPasswordTouchableOpacity>
        <Spacer.Column numberOfSpaces={8} />
        <ButtonPrimary
          title={_(t`Se connecter`)}
          onPress={handleSignin}
          disabled={shouldDisableLoginButton}
        />
      </BottomCard>
    </SafeContainer>
  )
}

const StyledInput = styled.View({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '100%',
  maxWidth: getSpacing(125),
})

const StyledInline = styled.View({
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-around',
  alignItems: 'center',
  maxWidth: getSpacing(125),
})

const StyledForgottenPasswordTouchableOpacity = styled.TouchableOpacity({
  flexDirection: 'row',
  justifyContent: 'flex-end',
  width: '100%',
  maxWidth: getSpacing(125),
})
