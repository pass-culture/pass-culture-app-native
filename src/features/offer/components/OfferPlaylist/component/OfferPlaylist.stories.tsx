import { ComponentMeta, ComponentStory } from '@storybook/react'
import React from 'react'
import styled from 'styled-components/native'

import { HitOfferWithArtistAndEan } from 'features/offer/components/OfferPlaylist/api/fetchOffersByArtist'
import { OfferPlaylist } from 'features/offer/components/OfferPlaylist/component/OfferPlaylist'
import { PlaylistType } from 'features/offer/enums'
import {
  mockedAlgoliaOffersWithSameArtistResponse,
  moreHitsForSimilarOffersPlaylist,
} from 'libs/algolia/__mocks__/mockedAlgoliaResponse'
import { getDisplayPrice } from 'libs/parsers'
import { Offer } from 'shared/offer/types'

const meta: ComponentMeta<typeof OfferPlaylist> = {
  title: 'features/offer/OfferPlaylist',
  component: OfferPlaylist,
}
export default meta

const Template: ComponentStory<typeof OfferPlaylist> = (props) => <OfferPlaylist {...props} />

const renderItem = (props: {
  item: Offer | HitOfferWithArtistAndEan
  width: number
  height: number
  playlistType?: PlaylistType
}) => (
  <StyledOfferTile>
    <StyledImage
      source={{
        uri: 'https://img.freepik.com/photos-gratuite/produit-beaute-explosant-couleurs-vibrantes-ia-generatrice-poudre_188544-9687.jpg?size=626&ext=jpg&ga=GA1.1.953763522.1701356022',
      }}
      accessibilityLabel={props.item.offer.name}
    />
    <StyledOfferName>{props.item.offer.name}</StyledOfferName>
    <StyledOfferPrice>{getDisplayPrice(props.item.offer.prices)}</StyledOfferPrice>
  </StyledOfferTile>
)

export const SameArtistPlaylist = Template.bind({})
SameArtistPlaylist.args = {
  items: mockedAlgoliaOffersWithSameArtistResponse,
  title: 'Du même auteur',
  renderItem,
  itemHeight: 250,
  itemWidth: 200,
  playlistType: PlaylistType.SAME_ARTIST_PLAYLIST,
}

export const SimilarOfferPlaylist = Template.bind({})
SimilarOfferPlaylist.args = {
  items: moreHitsForSimilarOffersPlaylist,
  title: 'Ça peut aussi te plaire',
  renderItem,
  itemHeight: 250,
  itemWidth: 200,
  playlistType: PlaylistType.SAME_CATEGORY_SIMILAR_OFFERS,
}

const StyledOfferTile = styled.View({
  padding: '5px',
  margin: '5px',
  borderRadius: '5px',
  width: '200px',
})

const StyledImage = styled.Image({
  width: '100%',
  height: '200px',
  marginBottom: '10px',
  borderRadius: '8px',
})

const StyledOfferName = styled.Text({
  fontSize: '12px',
  fontWeight: 'bold',
})

const StyledOfferPrice = styled.Text({
  fontSize: '14px',
  color: '#6e6f74',
})
