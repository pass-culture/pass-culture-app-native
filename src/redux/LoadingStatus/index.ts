export * from './actions';
export { loadingStatusReducer } from './reducer';
