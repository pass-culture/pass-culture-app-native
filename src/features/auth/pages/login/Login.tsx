import { yupResolver } from '@hookform/resolvers/yup'
import { useNavigation, useRoute } from '@react-navigation/native'
import React, { FunctionComponent, memo, useCallback, useEffect, useState } from 'react'
import { useForm } from 'react-hook-form'
import { Keyboard } from 'react-native'
import styled from 'styled-components/native'

import { useSignIn } from 'features/auth/api/useSignIn'
import { AuthenticationButton } from 'features/auth/components/AuthenticationButton/AuthenticationButton'
import { SSOButtonBase } from 'features/auth/components/SSOButton/SSOButtonBase'
import { useSettingsContext } from 'features/auth/context/SettingsContext'
import { loginSchema } from 'features/auth/pages/login/schema/loginSchema'
import { SignInResponseFailure } from 'features/auth/types'
import {
  StepperOrigin,
  UseNavigationType,
  UseRouteType,
} from 'features/navigation/RootNavigator/types'
import { analytics } from 'libs/analytics'
import { useFeatureFlag } from 'libs/firebase/firestore/featureFlags/useFeatureFlag'
import { RemoteStoreFeatureFlags } from 'libs/firebase/firestore/types'
import { useSafeState } from 'libs/hooks'
import { captureMonitoringError } from 'libs/monitoring'
import { ReCaptchaError, ReCaptchaInternalError } from 'libs/recaptcha/errors'
import { ReCaptcha } from 'libs/recaptcha/ReCaptcha'
import { EmailInputController } from 'shared/forms/controllers/EmailInputController'
import { PasswordInputController } from 'shared/forms/controllers/PasswordInputController'
import { ButtonPrimary } from 'ui/components/buttons/ButtonPrimary'
import { ButtonTertiaryBlack } from 'ui/components/buttons/ButtonTertiaryBlack'
import { Form } from 'ui/components/Form'
import { SUGGESTION_DELAY_IN_MS } from 'ui/components/inputs/EmailInputWithSpellingHelp/useEmailSpellingHelp'
import { InputError } from 'ui/components/inputs/InputError'
import { SeparatorWithText } from 'ui/components/SeparatorWithText'
import { SNACK_BAR_TIME_OUT_LONG, useSnackBarContext } from 'ui/components/snackBar/SnackBarContext'
import { SecondaryPageWithBlurHeader } from 'ui/pages/SecondaryPageWithBlurHeader'
import { Key } from 'ui/svg/icons/Key'
import { Spacer, Typo } from 'ui/theme'
import { getHeadingAttrs } from 'ui/theme/typographyAttrs/getHeadingAttrs'

type LoginFormData = {
  email: string
  password: string
}

type Props = {
  doNotNavigateOnSigninSuccess?: boolean
}

export const Login: FunctionComponent<Props> = memo(function Login(props) {
  const enableGoogleSSO = useFeatureFlag(RemoteStoreFeatureFlags.WIP_ENABLE_GOOGLE_SSO)
  const { data: settings } = useSettingsContext()
  const { params } = useRoute<UseRouteType<'Login'>>()
  const { navigate } = useNavigation<UseNavigationType>()
  const { showInfoSnackBar, showErrorSnackBar } = useSnackBarContext()
  const [isDoingReCaptchaChallenge, setIsDoingReCaptchaChallenge] = useState(false)
  const isRecaptchaEnabled = settings?.isRecaptchaEnabled

  const {
    handleSubmit,
    control,
    watch,
    setError: setFormErrors,
    formState: { isValid },
  } = useForm<LoginFormData>({
    mode: 'all',
    resolver: yupResolver(loginSchema),
    defaultValues: { email: '', password: '' },
    delayError: SUGGESTION_DELAY_IN_MS,
  })
  const email = watch('email')

  const [errorMessage, setErrorMessage] = useSafeState<string | null>(null)

  useEffect(() => {
    if (params?.displayForcedLoginHelpMessage) {
      showInfoSnackBar({
        message:
          'Pour sécuriser ton pass Culture, tu dois confirmer tes identifiants tous les 30 jours.',
        timeout: SNACK_BAR_TIME_OUT_LONG,
      })
    }
  }, [params?.displayForcedLoginHelpMessage, showInfoSnackBar])

  const handleSigninFailure = useCallback(
    (response: SignInResponseFailure) => {
      const failureCode = response.content?.code
      if (failureCode === 'SSO_EMAIL_NOT_FOUND') {
        navigate('SignupForm', {
          accountCreationToken: response.content?.accountCreationToken,
          email: response.content?.email,
          from: StepperOrigin.LOGIN,
        })
      } else if (
        failureCode &&
        [
          'DUPLICATE_GOOGLE_ACCOUNT',
          'SSO_ACCOUNT_DELETED',
          'SSO_ACCOUNT_ANONYMIZED',
          'SSO_EMAIL_NOT_VALIDATED',
        ].includes(failureCode)
      ) {
        showErrorSnackBar({
          message:
            'Ton compte Google semble ne pas être valide. Pour pouvoir te connecter, confirme d’abord ton adresse e-mail Google.',
          timeout: SNACK_BAR_TIME_OUT_LONG,
        })
      } else if (failureCode === 'EMAIL_NOT_VALIDATED') {
        navigate('SignupConfirmationEmailSent', { email })
      } else if (failureCode === 'ACCOUNT_DELETED') {
        setFormErrors('email', { message: 'Cette adresse e-mail est liée à un compte supprimé' })
      } else if (failureCode === 'NETWORK_REQUEST_FAILED') {
        setErrorMessage('Erreur réseau. Tu peux réessayer une fois la connexion réétablie')
      } else if (response.statusCode === 429 || failureCode === 'TOO_MANY_ATTEMPTS') {
        setErrorMessage('Nombre de tentatives dépassé. Réessaye dans 1 minute')
      } else {
        setErrorMessage('E-mail ou mot de passe incorrect')
      }
    },
    [showErrorSnackBar, navigate, email, setFormErrors, setErrorMessage]
  )

  const { mutate: signIn, isLoading } = useSignIn({
    doNotNavigateOnSigninSuccess: props.doNotNavigateOnSigninSuccess,
    params,
    setErrorMessage,
    onFailure: handleSigninFailure,
  })

  const shouldDisableLoginButton = !isValid || isLoading || isDoingReCaptchaChallenge

  const openReCaptchaChallenge = useCallback(() => {
    setIsDoingReCaptchaChallenge(true)
    setErrorMessage(null)
  }, [setErrorMessage])

  const onReCaptchaClose = useCallback(() => {
    setIsDoingReCaptchaChallenge(false)
  }, [])

  const onReCaptchaError = useCallback(
    (errorCode: ReCaptchaError, error?: string | undefined) => {
      setIsDoingReCaptchaChallenge(false)
      if (errorCode === ReCaptchaInternalError.NetworkError) {
        setErrorMessage(
          'Un problème est survenu, vérifie ta connexion internet avant de rééssayer.'
        )
      } else {
        setErrorMessage('Un problème est survenu, réessaie plus tard.')
        captureMonitoringError(`${errorCode} ${error}`, 'LoginOnRecaptchaError')
      }
    },
    [setErrorMessage]
  )

  const onReCaptchaExpire = useCallback(() => {
    setIsDoingReCaptchaChallenge(false)
    setErrorMessage('Le token reCAPTCHA a expiré, tu peux réessayer.')
  }, [setErrorMessage])

  const onReCaptchaSuccess = useCallback(
    (token: string) => {
      setIsDoingReCaptchaChallenge(false)
      handleSubmit((data) => signIn({ identifier: data.email, password: data.password, token }))()
    },
    [handleSubmit, signIn]
  )

  const onSubmit = useCallback(
    async (data: LoginFormData) => {
      if (!shouldDisableLoginButton) {
        setErrorMessage('')
        Keyboard.dismiss()
        isRecaptchaEnabled
          ? openReCaptchaChallenge()
          : signIn({ identifier: data.email, password: data.password })
      }
    },
    [shouldDisableLoginButton, isRecaptchaEnabled, setErrorMessage, openReCaptchaChallenge, signIn]
  )

  const onForgottenPasswordClick = useCallback(() => {
    navigate('ForgottenPassword')
  }, [navigate])

  const onLogSignUpAnalytics = useCallback(() => {
    analytics.logSignUpClicked({ from: 'login' })
  }, [])

  return (
    <React.Fragment>
      {!!isRecaptchaEnabled && (
        <ReCaptcha
          onClose={onReCaptchaClose}
          onError={onReCaptchaError}
          onExpire={onReCaptchaExpire}
          onSuccess={onReCaptchaSuccess}
          isVisible={isDoingReCaptchaChallenge}
        />
      )}
      <SecondaryPageWithBlurHeader headerTitle="Connexion" shouldDisplayBackButton>
        <Spacer.Column numberOfSpaces={6} />
        <Typo.Title3 {...getHeadingAttrs(1)}>Connecte-toi</Typo.Title3>
        <Spacer.Column numberOfSpaces={2} />
        <Form.MaxWidth>
          <InputError
            visible={!!errorMessage}
            messageId={errorMessage}
            numberOfSpacesTop={5}
            centered
          />
          <Spacer.Column numberOfSpaces={7} />
          <EmailInputController name="email" control={control} isRequiredField />
          <Spacer.Column numberOfSpaces={6} />
          <PasswordInputController
            name="password"
            control={control}
            onSubmitEditing={handleSubmit(onSubmit)}
            isRequiredField
          />
          <Spacer.Column numberOfSpaces={5} />
          <ButtonContainer>
            <ButtonTertiaryBlack
              wording="Mot de passe oublié&nbsp;?"
              onPress={onForgottenPasswordClick}
              icon={Key}
              inline
            />
          </ButtonContainer>
          <Spacer.Column numberOfSpaces={8} />
          <ButtonPrimary
            wording="Se connecter"
            onPress={handleSubmit(onSubmit)}
            disabled={shouldDisableLoginButton}
          />
        </Form.MaxWidth>
        {enableGoogleSSO ? (
          <React.Fragment>
            <Spacer.Column numberOfSpaces={4} />
            <StyledSeparatorWithText label="ou" />
            <Spacer.Column numberOfSpaces={4} />
            <SSOButtonBase type="login" onSuccess={signIn} />
            <Spacer.Column numberOfSpaces={10} />
          </React.Fragment>
        ) : (
          <Spacer.Column numberOfSpaces={8} />
        )}
        <SignUpButton onAdditionalPress={onLogSignUpAnalytics} />
      </SecondaryPageWithBlurHeader>
    </React.Fragment>
  )
})

const ButtonContainer = styled.View(({ theme }) => ({
  flexDirection: 'row',
  width: '100%',
  maxWidth: theme.buttons.maxWidth,
}))

const SignUpButton = styled(AuthenticationButton).attrs(({ theme }) => ({
  linkColor: theme.colors.secondary,
  type: 'signup',
}))``

const StyledSeparatorWithText = styled(SeparatorWithText).attrs(({ theme }) => ({
  backgroundColor: theme.colors.greyMedium,
}))``
