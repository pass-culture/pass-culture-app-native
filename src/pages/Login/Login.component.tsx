import React, { FunctionComponent } from 'react';
import { Button, View, Text, ActivityIndicator, StyleSheet } from 'react-native';
import { useDispatch, useSelector } from 'react-redux'; // @redux
import { StackNavigationProp } from '@react-navigation/stack';
import { RouteProp } from '@react-navigation/native';
import { t } from '@lingui/macro';
import { i18n } from '../../lib/i18n';

import { CheatCodesButton } from '../../atoms/CheatCodesButton';
import { RootStackParamList } from '../../navigation';
import { isLoadingSelector } from '../../redux/LoadingStatus/selectors'; // @redux
import { startLoading, finishLoading } from '../../redux/LoadingStatus'; // @redux
import { setToken } from '../../redux/Authentication'; // @redux
import { tokenSelector } from '../../redux/Authentication/selectors'; // @redux-persist-sensitive

type LoginScreenNavigationProp = StackNavigationProp<RootStackParamList, 'Login'>;
type LoginScreenRouteProp = RouteProp<RootStackParamList, 'Login'>;

const styles = StyleSheet.create({
  container: { flex: 1, alignItems: 'center', justifyContent: 'center' },
});
const LOADER_NAME = 'demoLoader';

type Props = {
  navigation: LoginScreenNavigationProp;
  route: LoginScreenRouteProp;
};
export const Login: FunctionComponent<Props> = ({ navigation, route }) => {
  const dispatch = useDispatch();
  const isLoading = useSelector(isLoadingSelector(LOADER_NAME));
  const authenticationToken = useSelector(tokenSelector);

  const goToHomePage = (): void => {
    navigation.navigate('Home');
  };

  const toggleLoader = () => {
    if (!isLoading) {
      dispatch(startLoading(LOADER_NAME));
    } else {
      dispatch(finishLoading(LOADER_NAME));
    }
  };

  const generateAuthenticationToken = () => {
    const randomToken = `token-${Math.floor(Math.random() * 100 + 1)}`;
    dispatch(setToken(randomToken));
  };

  return (
    <View style={styles.container}>
      <Text>{i18n._(/*i18n: The title of the login page */ t`Login page`)}</Text>
      <Button
        testID="homepageButton"
        title={i18n._(/*i18n: The title of button to go to the home page */ t`Go to Home Page`)}
        onPress={goToHomePage}
      />
      <CheatCodesButton navigation={navigation} />
      <Text>{i18n._(/*i18n: The title of button to test redux */ t`Redux test`)}</Text>
      <Button
        testID="tokenButton"
        title={i18n._(
          /*i18n: The title of button to generate a random token */ t`Generate a random token`
        )}
        onPress={generateAuthenticationToken}
      />
      <Text>{authenticationToken}</Text>
      <Button
        testID="loaderButton"
        title={i18n._(/*i18n: The title of button to toggle loader */ t`Toggle selector`)}
        onPress={toggleLoader}
      />
      {isLoading && <ActivityIndicator testID="demoLoader" />}
      <Text>{route.params && route.params.userId}</Text>
    </View>
  );
};
