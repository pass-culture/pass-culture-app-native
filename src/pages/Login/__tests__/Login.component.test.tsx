import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';
import configureStore from 'redux-mock-store'; // @redux
import { Provider } from 'react-redux'; // @redux
import { startLoading, finishLoading } from '../../../redux/LoadingStatus'; // @redux
import { Login } from '../Login.component';

const mockStore = configureStore([]);
const initialState = { loadingStatus: {}, authentication: {} };

describe('Login component', () => {
  const navigation = {
    navigate: jest.fn(),
  } as any; // eslint-disable-line @typescript-eslint/no-explicit-any
  const route = {
    params: { userId: 'testId' },
  } as any; // eslint-disable-line @typescript-eslint/no-explicit-any

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const login = render(
      <Provider store={store}>
        <Login navigation={navigation} route={route} />
      </Provider>
    );
    expect(login).toMatchSnapshot();
  });

  it('should have a button to go to the Home page', () => {
    const store = mockStore(initialState);
    const login = render(
      <Provider store={store}>
        <Login navigation={navigation} route={route} />
      </Provider>
    );

    const HomeButton = login.getByTestId('homepageButton');
    fireEvent.press(HomeButton);
    expect(navigation.navigate).toHaveBeenCalledWith('Home');
  });

  it('should trigger the loader', () => {
    const store = mockStore(initialState);
    const login = render(
      <Provider store={store}>
        <Login navigation={navigation} route={route} />
      </Provider>
    );

    expect(() => login.getByTestId('demoLoader')).toThrowError();

    const LoaderButton = login.getByTestId('loaderButton');
    fireEvent.press(LoaderButton);

    expect(store.getActions()).toEqual([startLoading('demoLoader')]);
  });

  it('should stop the loader', () => {
    const store = mockStore({
      ...initialState,
      loadingStatus: { demoLoader: { isLoading: true } },
    });
    const login = render(
      <Provider store={store}>
        <Login navigation={navigation} route={route} />
      </Provider>
    );

    expect(login.getByTestId('demoLoader')).toBeDefined();

    const LoaderButton = login.getByTestId('loaderButton');
    fireEvent.press(LoaderButton);

    expect(store.getActions()).toEqual([finishLoading('demoLoader')]);
  });
});
