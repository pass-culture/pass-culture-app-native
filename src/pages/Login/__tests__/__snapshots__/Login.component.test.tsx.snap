// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Login page
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="homepageButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go to Home Page
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        CheatCodes
      </Text>
    </View>
  </View>
  <Text>
    Redux test
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="tokenButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Generate a random token
      </Text>
    </View>
  </View>
  <Text />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="loaderButton"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Toggle selector
      </Text>
    </View>
  </View>
  <Text>
    testId
  </Text>
</View>
`;
